spec/migrations/20231016094900_microsecond_timestamp_msql_asg_update_spec.rb

```diff
@@ -0,0 +1,42 @@
+require 'spec_helper'
+require 'migrations/helpers/migration_shared_context'
+
+RSpec.describe 'migration to enable microsecond precision on asg last updated table', isolation: :truncation do
+  include_context 'migration' do
+    let(:migration_filename) { '20231016094900_microsecond_timestamp_msql_asg_update.rb' }
+    let(:ds) do
+      db[:asg_timestamps].with_extend do
+        def supports_timestamp_usecs?
+          true
+        end
+
+        def timestamp_precision
+          6
+        end
+      end
+    end
+  end
+
+  describe 'asg_timestamps table' do
+    it 'the last_update column handles sub-second time' do
+      initial_time = DateTime.new(2010)
```

Comment:
On mysql this is the case, but on postgres, fractional seconds are supported. We wanted to have a test that worked in both situations:

1. Before test - do the sub-seconds match what we generated before inserting + after?
2. After tests:
- does the original sub-second remain unchanged after migrating?
- are we able to store sub-second times and retrieve them to be the same as what we inserted?

Reply:
I think it's better to have an explicit check for the database type here as MySQL and Postgres are behaving differently. And the Sequel migration is also only done for one database type.

When inserting a timestamp without fractional seconds, one cannot tell afterwards if they are missing because of the database or the initial value...

My proposal is to insert a timestamp with fractional seconds and then check if the stored value:
- does not contain fractional seconds (on MySQL), or
- equals the original value (on Postgres).

And after applying the migration, both database types should have the same behavior.

```
dt_without_fractional_seconds = DateTime.new(2001, 2, 3, 4, 5, 6)
dt_with_fractional_seconds = dt_without_fractional_seconds.advance(seconds: 0.123456)

ds.insert(id: 1, last_update: dt_with_fractional_seconds)
t1 = ds.first(id: 1)

if db.database_type == :mysql
  expect(t1[:last_update].to_datetime.subsec).to eq(0)
  expect(t1[:last_update].to_datetime.rfc3339(6)).to eq(dt_without_fractional_seconds.rfc3339(6))
else
  expect(t1[:last_update].to_datetime.subsec).not_to eq(0)
  expect(t1[:last_update].to_datetime.rfc3339(6)).to eq(dt_with_fractional_seconds.rfc3339(6))
end

# Change TIMESTAMP to TIMESTAMP(6)
expect { Sequel::Migrator.run(db, migration_to_test, allow_missing_migration_files: true) }.not_to raise_error

# the migration shouldn't add accuracy to previously inserted values
t1_post_migration = ds.first(id: 1)
expect(t1_post_migration[:last_update].to_datetime.rfc3339(6)).to eq(t1[:last_update].to_datetime.rfc3339(6))

# but new data should retain microsecond accuracy
ds.insert(id: 2, last_update: dt_with_fractional_seconds)
t2 = ds.first(id: 2)

expect(t2[:last_update].to_datetime.subsec).not_to eq(0)
expect(t2[:last_update].to_datetime.rfc3339(6)).to eq(dt_with_fractional_seconds.rfc3339(6))
```